### ISSUE TITLE
After logging in with interactive the application does not grab the response.

### ISSUE THREAD

**Issue by asusralis** [AzureAD/microsoft-authentication-library-for-dotnet#1477]
**Which Version of MSAL are you using ?**
4.5.1

**Platform**
Xamarin.Forms - testing on android.

**What authentication flow has the issue?**
* Desktop / Mobile
    * [x] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App
    * [ ] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

**Is this a new or existing app?**
New app.

```
<activity android:name="microsoft.identity.client.BrowserTabActivity">
      <intent-filter>
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="msal[removed]" android:host="auth" />
      </intent-filter>
    </activity>

        protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
        {
            base.OnActivityResult(requestCode, resultCode, data);

            AuthenticationContinuationHelper.SetAuthenticationContinuationEventArgs(requestCode, resultCode, data);
        }

...

IPublicClientApplication client = PublicClientApplicationBuilder.Create(B2CConstants.ClientID)
                .WithB2CAuthority(B2CConstants.AuthoritySignInSignUp)
                .WithIosKeychainSecurityGroup(B2CConstants.IOSKeyChainGroup)
                .WithRedirectUri("https://login.microsoftonline.com/tfp/oauth2/nativeclient")
                .Build();

                AuthenticationResult result =  await client.AcquireTokenInteractive(B2CConstants.Scopes)
                .WithParentActivityOrWindow(App.Activity)
                //.WithUseEmbeddedWebView(true)
                .ExecuteAsync();
```

I am able to log into the form it brings when this is called, but the application does not grab the response after the log in is successful. If I manually go back it will say it was canceled by the user. The response's url is something like: "?state = very long token" so I assumed that part is working.

When I set "WithUseEmbeddedWebView" to true the application comes back after I log in but I get this exception which I don't quite understand:
```
Microsoft.Identity.Client.MsalServiceException: 'AADB2C90079: Clients must send a client_secret when redeeming a confidential grant.
```

**Comment by jennyf19** [comment]
@asusralis Can you try [this ](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Client-Applications#redirect-uri-for-public-client-applications)as the redirect uri? Make sure to register it in the portal:
`.WithRedirectUri($"msal{ClientID}://auth")`
and then make sure in the Android manifest, you have the same clientId:
`<data android:scheme="msal[clientId]" android:host="auth" />`

Also, which identity provider are you using when you sign in? Google, FB, B2C Local Account, etc...? 

One last thing, in the portal, did you create a public client? Seems to be that is what you are using on mobile?
![image](https://user-images.githubusercontent.com/19942418/67643310-56d37f80-f8d3-11e9-8871-6af1aa2fd253.png)
thanks.

**Comment by asusralis** [comment]
Thank you for the response! I missed the page you sent me. I removed the redirect method and it worked! However, for some reason it did not send an access token. 

A little info: I'm signing in with both Google and/or username/password. Does native client mean public? If so, yes. 

I have my api (asp.net) registered as a web app and my Xamarin.Forms application as a native client. I'm trying to get an access on my Xamarin.Forms app so I can authenticate my REST calls to my api, but for some reason it doesn't want to return an access token. Would there be any reason to why the access token is null? Thanks again for the help!

**Comment by jmprieur** [comment]
@asusralis : yes, native client means public client application, that is desktop/mobile apps.

Did you register in the portal that your Xamarin.Forms application wants to call your Web API? which scope do you use for your Web API?

**Comment by asusralis** [comment]
Would that be under API access or Published scopes? In API access there are two permitted scopes - openid and offline_access. The Xamarin.Forms application is the same.

For published scopes the web api has "user_impersonation - Access this app on behalf of the signed-in user - ...API/user_"

In the Xamarin.Forms' Published scopes it says, "Web API has not been included," but I don't seem to see an option to change that.

/API/user_impersonation |  
-- | --
 



/API/user_impersonation |  
-- | --

**Comment by jmprieur** [comment]
that would be in the app registration for your Xamarin app, under API access. 

In the published scopes for your Web API, It should mention the scope to use.

**Comment by jmprieur** [comment]
@asusralis, this would be in the API access part for your Xamarin app. you need to declare that your Xamarin app whants to access your Web API.

**Comment by asusralis** [comment]
I see! But I'm still confused - why would this stop me for getting an access token from the interactive method called in Xamarin.Forms? It wouldn't generate one if I wasn't using a web api?

**Comment by asusralis** [comment]
It still seems the result does not have the access token: https://i.imgur.com/Q2dQTnX.png

**Comment by jmprieur** [comment]
do you mind sharing with me B2CConstant.Scopes ?

**Comment by asusralis** [comment]
I declared them like this: 
```
        public static string[] Scopes = {
            "https://xxx.onmicrosoft.com/Xamarin.Forms/read",
            "https://xxx.onmicrosoft.com/Xamarin.Forms/openid",
            "https://xxx.onmicrosoft.com/Xamarin.Forms/offline_access"
        };
```

I saw multiple ways to do them.

**Comment by jmprieur** [comment]
@asusralis : you don't need any of these.
what you need is "yourWebApi/user_impersonation" (from what I read above)

Something like that:
        public static string[] Scopes = { "https://fabrikamb2c.onmicrosoft.com/helloapi/demo.read" };

See
https://github.com/Azure-Samples/active-directory-b2c-xamarin-native/blob/5b5446824967a786a1decc789c0580e1aa64e8ab/UserDetailsClient/UserDetailsClient.Core/Features/LogOn/B2CConstants.cs#L13

BTW, did you have a look at the following sample and its README.md ?
https://github.com/Azure-Samples/active-directory-b2c-xamarin-native

**Comment by asusralis** [comment]
Wow, that worked! Sorry, I was *really* confused by the scopes. I wasn't sure what the "demo.read" meant, but I see where it corresponds to now. Thanks for the help!

**Comment by jmprieur** [comment]
I'm glad you made it work, @asusralis 

Can you please help us improving our docs and guidance by answering the following questions?
What could we have done better so that you had not been confused in the first place?
Where did you get the idea of using the scopes you had used?
What was the haha moment where you understood?
Ideally what would have helped you getting straight to the solution?

**Comment by asusralis** [comment]
I think I was just confused about how to use the scopes and needing to declare that my client wanted to access my api, but I don't think my ignorance came from this repo's documents. I used the scopes I had from a docs.microsoft.com webpage about azure AD. 

My needs from AD are fairly simple but AD - and this repo - seems quite large, so there was just a lot of reading I had to go over. For the time spent on this I ended up needing to write only around 20 lines of code for it to work.
